Userland/Libraries/LibDeviceTree/DeviceTree.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2024, Leon Albrecht <leon.a@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+#include <AK/Endian.h>
+#include <AK/Function.h>
+#include <AK/HashMap.h>
+#include <AK/IterationDecision.h>
+#include <AK/MemoryStream.h>
+#include <AK/Optional.h>
+#include <AK/Span.h>
+
+namespace DeviceTree {
+
+struct DeviceTreeProperty {
+    ReadonlyBytes raw_data;
+
+    size_t size() const { return raw_data.size(); }
+
+    StringView as_string() const { return StringView(raw_data.data(), raw_data.size() - 1); }
+    Vector<StringView> as_strings() const { return as_string().split_view('\0'); }
+
+    // Note: as<T> does not convert endianness, so all structures passed in
+    //       should use BigEndian<T>s for their members and keep ordering in mind
+    // Note: The Integral variant does convert endianness, so no need to pass in BigEndian<T>s
+    template<typename T>
+    T as() const
+    {
+        VERIFY(raw_data.size() == sizeof(T));
+        T value;
+        __builtin_memcpy(&value, raw_data.data(), sizeof(T));
+        return value;
+    }
+
+    template<typename T>
+    requires(alignof(T) <= 4 && !IsIntegral<T>)
+    T const& as() const
+    {
+        return *reinterpret_cast<T const*>(raw_data.data());
+    }
+
+    template<Integral I>
+    I as() const
+    {
+        VERIFY(raw_data.size() == sizeof(I));
+        BigEndian<I> value;
+        __builtin_memcpy(&value, raw_data.data(), sizeof(I));
+        return value;
+    }
+
+    template<typename T>
+    ErrorOr<void> for_each_in_array_of(Function<ErrorOr<IterationDecision>(T const&)> function) const
+    {
+        VERIFY(raw_data.size() % sizeof(T) == 0);
+        size_t count = raw_data.size() / sizeof(T);
+        size_t offset = 0;
+        for (size_t i = 0; i < count; ++i, offset += sizeof(T)) {
+            auto sub_property = DeviceTreeProperty { raw_data.slice(offset, sizeof(T)) };
+            auto result = function(sub_property.as<T>());
+            if (result.is_error())
+                return result;
+            if (result.value() == IterationDecision::Break)
+                break;
+        }
+        return {};
+    }
+
+    FixedMemoryStream as_stream() const { return FixedMemoryStream { raw_data }; }
+};
+
+class DeviceTreeNodeView {
+public:
+    bool has_property(StringView prop) const { return m_properties.contains(prop); }
+    bool has_child(StringView prop) const { return m_children.contains(prop); }
```

Comment:
This parameter should probably be called `child` or `name` (as you called it in `get_child`)